include/gsl/gsl-lite.hpp

```diff
@@ -1540,32 +1540,22 @@ class not_null
     }
 #endif // gsl_HAVE( RVALUE_REFERENCE )
 
-#if gsl_HAVE( FUNCTION_REF_QUALIFIER )
-# if gsl_CONFIG( NOT_NULL_TRANSPARENT_GET )
-    gsl_api gsl_constexpr14 element_type &  get() const &  { return checked_ptr().get(); }
-    gsl_api gsl_constexpr14 element_type && get() &&       { return std::move(checked_ptr()).get(); }
-# else
-    gsl_api gsl_constexpr14 get_result_t get() const { return checked_ptr(); }
-# endif
+#if gsl_CONFIG( NOT_NULL_TRANSPARENT_GET )
+    gsl_api gsl_constexpr14 element_type* get() const { return checked_ptr().get(); }
+#else
+    gsl_api gsl_constexpr14 get_result_t  get() const { return checked_ptr(); }
+#endif
 
+#if gsl_HAVE( FUNCTION_REF_QUALIFIER )
```

Comment:
I'd be very unhappy with that because then our default configuration wouldn't support moving from `not_null<unique_ptr<T>>`.

Ideally, we could have
```c++
    gsl_api gsl_constexpr14 operator T const &() const & { return checked_ptr(); }
    gsl_api gsl_constexpr14 operator T &&     () &&      { return std::move(checked_ptr()); }
```

so we would avoid a copy but still permit a move. However, this makes the following code ambiguous:

```c++
not_null<unique_ptr<T>> np = ...;
unique_ptr<T> p = std::move(np);
```

However, I think having the implicit conversion operator return a value is justifiable in C++11 because move semantics alleviate the cost of moving the returned value around.

The only case where converting to `T const &` would save you a copy is when interacting with code like this, which I'd hope to be rare:

```c++
// Maybe we want a copy, maybe we don't, so take the arg by const ref.
// (If we knew we needed a copy, we'd simply take the arg by value;
//  if we only wanted to borrow the pointer we'd just take a plain `int const *` instead.)
void foo( std::shared_ptr<int> const & );
```

I think not supporting moving from non-copyable smart pointers is too high a price to pay for this.